File: solutions/prolog/knapsack/1/knapsack.pl
maximum_value([], _, 0).
maximum_value(Items, Capacity, Value):-
    maximum_value(Items, Capacity, 0, Value).
maximum_value([], _, Value, Value):-
    Value > 0.
maximum_value([item(W,V)|Items], Capacity, Acc, Value):-
    Capacity >= W,
    NewCapacity is Capacity - W,
    Acc1 is V + Acc,
    maximum_value(Items, NewCapacity, Acc1, Value).
maximum_value([_|Items], Capacity, Acc, Value):-
    maximum_value(Items, Capacity, Acc, Value).

File: solutions/java/knapsack/1/src/main/java/Knapsack.java
import java.util.List;

class Knapsack {

    int maximumValue(int maximumWeight, List<Item> items) {
        int[] sack = new int[maximumWeight + 1];
        for(Item item: items) {
            for(int w = maximumWeight; w >= item.weight; w--) {
                sack[w] = Math.max(sack[w], sack[w - item.weight] + item.value);
            }
        }
        return sack[maximumWeight];
    }

}